### ISSUE TITLE
server_environment_ir_config_parameter security purpose

### ISSUE THREAD

**Issue by bealdav** [OCA/server-env#25]
Hi all,

cc @sbidoul @guewen @yvaucher @gurneyalex 

I dived into this module

https://github.com/OCA/server-env/blob/12.0/server_environment_ir_config_parameter

I really appreciate to have a way to combine server_env module without to have an UI to build (then a UI fallback). Thanks to the authors

But I have a question about security or confidentiality of informations

A parameter can be a credential (all that we have in server_env keys), and I ask myself why values contained in server_env files
are always written in ir.config_parameter instead of only used as substitution.

https://github.com/OCA/server-env/blob/12.0/server_environment_ir_config_parameter/models/ir_config_parameter.py#L30

IMO, i thought that UI was for demo purposes not for real production storage informations place.

These values are also in backup, and that may imply to rise questions in gdpr context I suppose.

I think we could have an optional behaviour to prevent to store data in db.
What do you think?

Maybe I missed something, so please make me take the right way.

Thanks

**Comment by sbidoul** [comment]
I wrote this module. The initial goal was to override keys in arbitrary modules that store their config in ir.config.parameter. I think my first use case was for `web_environment_ribbon`.

So in those use cases, this module is not less nor more secure than the original that stores it's config in database.

I chose to store it in database for maximum compatibility. But I would be fine with a substitution approach. To be verified if there are no place that read from ir.config.parameters using another method than get_param.

**Comment by bealdav** [comment]
OK thanks.
Another PR in my todo list, then

**Comment by legalsylvain** [comment]
Hi. @bealdav. Did you worked on that point ? I'm using this module and I face the same problem. 
Also, the UI has a end user behaviour that is not intuitive : 

- We have a production database with some config_parameter written in the odoo.cfg files like ``my_key = my_production_value``
- make a dump, and restore the database for test / debug / spare / developpement purpose 
- launch the database in the test environment.
A) going in the ir.config_parameter view, the user will see the ``my_production_value``
B) when "consuming" with ``get_param()`` function, the value in db will be overwritten. 
C) going in the ir.config_parameter view, the user will see the ``my_test_value``

the A situation should not occures.

My proposal : 

- remove the line that writes [in the db](https://github.com/OCA/server-env/blob/12.0/server_environment_ir_config_parameter/models/ir_config_parameter.py#L30).
- add a field ``environment_value``, computed.
- display the field if set.

Pro:
-> fixes security issues. developer / tester will not have access to the production values;
-> more intuitive in the UI. (case A will not occures)
Cons:
-> if a core module, or an OCA modules makes a ``parameter.value`` instead of the ``parameterObj.get_param("key")``, we will introduces bugs.

My proposal is to makes the changes in stable version (V12 for instance), because such implementation should not exist and it's quite bad to have security issue, just to prevent hypothetical bugs to be raised, due to bad implementation.

What do you think ?

CC : @sbidoul, @rvalyi, @fcayre  

Note : I made an improvment in this module. Review welcome : https://github.com/OCA/server-env/pull/67

**Comment by sbidoul** [comment]
@legalsylvain my comment [above](https://github.com/OCA/server-env/issues/25#issuecomment-509670494) still applies :)

**Comment by legalsylvain** [comment]
Hi @sbidoul. Thanks for your answer. 
My question was about your answer "_To be verified if there are no place that read from ir.config.parameters using another method than get_param._"
I have no time to check the 300 odoo modules and the 200 OCA repo to verify if there a place where a bad ``read`` is done. (and in fact, I don't know if it's possible to ``rgrep`` such case).

So, if you're ok (as you are the author of this module), I'll do a PR, once the previous one #67 is merged.

(or I can include in #67, but I know you like the "_One PR, One change_" philosophy ;-)

thanks !

**Comment by bealdav** [comment]
Hi @legalsylvain no sorry not worked on it.
I like your proposal. I vote for it.

**Comment by sbidoul** [comment]
> (or I can include in #67, but I know you like the "One PR, One change" philosophy ;-)

better separate them as they are unrelated and #67 seems to be almost ready

> I have no time to check the 300 odoo modules and the 200 OCA repo to verify if there a place where a bad read is done. (and in fact, I don't know if it's possible to rgrep such case).

> So, if you're ok

Honestly I just don't know if there is a real risk at doing that or if the risk I perceived when I authored the module has been mitigated since. Maybe you could log an error if read is called (unless a specific context key is set) so we at least have a way to spot such cases ?
